File: advent14/14a.py
import sys
import re

class Machine:
    def __init__(self):
        self.memory = {}
        self.mask_mask = 0
        self.mask_bits = 0

    def set_mask(self, mask):
        self.mask_mask = 0
        self.mask_bits = 0
        for i, ch in enumerate(reversed(mask)):
            if ch != "X":
                self.mask_mask |= 1 << i
                if ch == "1":
                    self.mask_bits |= 1 << i

    def set_value(self, addr, val):
        self.memory[addr] = (val & ~self.mask_mask) | self.mask_bits

    def sum_values(self):
        return sum(self.memory.values())

machine = Machine()

for line in sys.stdin:
    if line.startswith("mem"):
        addr, val = re.findall(r"mem\[(\d+)\] = (\d+)", line)[0]
        machine.set_value(int(addr), int(val))
    elif line.startswith("mask"):
        mask = line.split()[-1]
        machine.set_mask(mask)

print(machine.sum_values())

File: advent14/14a.cpp
#include <iostream>
#include <cstdio>
#include <cinttypes>
#include <array>
#include <unordered_map>

/* NOTE:
    This was a good try but as C++ has no BigInt, it cannot sum
    all of the values in memory without overflowing. So I rewrote in in Python.
*/

// To build:
//  g++ 14a.cpp -std=c++2a -o 14a
//  cat input.txt | ./14a

using Addr = uint64_t;
using Value = uint64_t;

class Machine {
public:
    void SetMask(const std::string& mask) {
        Value m = 0;
        Value b = 0;
        size_t i = 0;

        for (auto it = mask.crbegin(); it != mask.crend(); ++it) {
            if (*it != 'X') {
                m |= (1 << i);
                if (*it == '1') {
                    b |= (1 << i);
                }
            }

            ++i;
        }

        this->maskMask = m;
        this->maskBits = b;
    }

    void SetMem(Addr addr, Value value) {
       Value newValue = this->maskBits | (value & ~this->maskMask);
       this->memory[addr] = newValue;
    }

    Value SumValues() const {
        Value sum = 0;
        for (auto const& [addr, val] : this->memory) {
            sum += val;
        }
        return sum;
    }
private:
    std::unordered_map<Addr, Value> memory;
    Value maskMask = 0;
    Value maskBits = 0;
};

int main(int argc, char** argv) {
    Machine m;
    std::string line;

    while (std::getline(std::cin, line)) {
        Addr addr;
        Value value;
        std::array<char, 40> mask;

        if (std::sscanf(line.c_str(), "mask = %s", mask.data()) == 1) {
            m.SetMask(mask.data());
        }
        else if (std::sscanf(line.c_str(), "mem[%" SCNu64 "] = %" SCNu64, &addr, &value) == 2) {
            m.SetMem(addr, value);
        }
    }

    std::cout << m.SumValues() << std::endl;

    return 0;
}